### PR TITLE
handle empty response

### PR DESCRIPTION
**Pull request by garasubo**
When the server sends RST packet without any contents, the incoming
thread will be shutdown with IO error. But outbound thread keeps alive
because it doesn't recognize incoming thread termination.

In this commit, I added channels to notify incoming thread termination
to outbound thread and outbound thread termination to incoming thread.

### CODE REVIEW

File: src/relay.rs [hunk 1]
```diff
@@ -12,10 +12,6 @@ use crate::thread::spawn_thread;
 
 #[derive(Debug)]
 pub struct RelayHandle {
-    /// client address
-    client_addr: SocketAddr,
-    /// server address
-    server_addr: SocketAddr,
```
Comment (legokichi) [inline src/relay.rs:18]: unused fields

File: src/relay.rs [hunk 2]
```diff
@@ -121,7 +130,16 @@ fn spawn_relay_half(
                 return Ok(());
             }
             Ok(size) => trace!("{}: {} ==> {}: {} bytes", name, src_addr, dst_addr, size),
-            Err(err) if err.kind() == K::WouldBlock || err.kind() == K::TimedOut => {}
+            Err(err) if err.kind() == K::WouldBlock || err.kind() == K::TimedOut => {
+                match another_rx.try_recv() {
```
[the remark this answers is not in the document]
Comment (garasubo) [reply src/relay.rs]: Sounds a good idea. Fixed

**Review by eldesh — CHANGES_REQUESTED** [review]
The shutdown_relay_by_connection_rest passes on the master branch before the merge.
It seems that the fixes are not being tested properly.

**Comment by garasubo** [comment]
@eldesh It seems all tests passed in this branch as well.
https://app.circleci.com/pipelines/github/Idein/gatekeeper/205/workflows/57a113cb-79b7-4022-a4ad-15831adae9d1
Is there any tests not running in this CI?

**Review by legokichi — DISMISSED** [review]
dante-server seems to pass RSTs transparently.
However, I think this PR is fine as long as the connection can be simply disconnected

**Comment by eldesh** [comment]
> It seems all tests passed in this branch as well.

Yes I see. It has passed all test cases.

But doesn't this PR fix bugs in the previous version?
If so, then the newly added test (`shutdown_relay_by_connection_rest`) should fail in the **master branch** (i.e, in the pre-fixed version), right?

**Comment by garasubo** [comment]
I see. Added Node.js based test. I firstly tried to add unit test in Rust, but it is a little bit difficult since there is no straight way to close TCP session with RST packet using std. So, I added a snippet, which was used to find this bug.

File: example.yml [hunk 3]
```diff
@@ -55,3 +55,25 @@
     port: Any
     protocol:
       Specif: Tcp
+- Allow:
```
Comment (eldesh) [inline example.yml:58]: Is this change as intended?

**Review by eldesh — DISMISSED** [review]
I have confirmed that the test fails in previous versions.
Looks good :+1:

File: example.yml (hunk 3 above)
Comment (garasubo) [reply example.yml]: No, removed

**Review by legokichi — APPROVED** [review]

**Pull request merged**